### PR TITLE
Santander Consumer Bank PDF import: Support removal transactions

### PR DESCRIPTION
**Pull request by auchri**

### CODE REVIEW

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java [hunk 1]
```diff
@@ -256,6 +256,31 @@ private void addAccountStatementTransaction()
                             t.setNote(TextUtil.replaceMultipleBlanks(v.get("note")).trim());
                         })
 
+                        .wrap(TransactionItem::new));
+
+        // 20.05.2023 20.05.2023 -1,00 33,01 Auszahlung auf ...
+        Block removalBlock = new Block(
+                        "^(?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) ([\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<amount>[\\.,\\d-]+) ([\\.,\\d-]+) (Auszahlung) (?<note>.*)");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java:263]: - Improve regulare expression
- Remove obsolet mapping

```suggestion
                        "^[\\d]{2}\\.[\\d]{2}\\.[\\d]{4} [\\d]{2}\\.[\\d]{2}\\.[\\d]{4} \\-[\\.,\\d]+ [\\.,\\d]+ Auszahlung .*$");
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java [hunk 2]
```diff
@@ -256,6 +256,31 @@ private void addAccountStatementTransaction()
...
+        removalBlock.set(new Transaction<AccountTransaction>()
+
+                        .subject(() -> {
+                            AccountTransaction entry = new AccountTransaction();
+                            entry.setType(AccountTransaction.Type.REMOVAL);
+                            return entry;
+                        })
+
+                        .section("date", "amount", "note")
+                        .match("^(?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) ([\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<amount>[-\\.,\\d-]+) ([\\.,\\d-]+) (Auszahlung) (?<note>.*)")
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java:274]: - Improve regulare expressions
```suggestion
                        .match("^(?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) [\\d]{2}\\.[\\d]{2}\\.[\\d]{4} \\-(?<amount>[\\.,\\d]+) [\\.,\\d]+ Auszahlung (?<note>.*)$")
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java [hunk 3]
```diff
@@ -256,6 +256,31 @@ private void addAccountStatementTransaction()
...
+                        .assign((t, v) -> {
+                            Map<String, String> context = type.getCurrentContext();
+
+                            t.setDateTime(asDate(v.get("date")));
+                            t.setAmount(asAmount(v.get("amount")));
+                            t.setCurrencyCode(asCurrencyCode(context.get("currency")));
+                            t.setNote(TextUtil.replaceMultipleBlanks(v.get("note")).trim());
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java:281]: You are already using static import of functions.
- Use static imports... see [Contributing rules --> Configure Eclipse](https://github.com/buchen/portfolio/blob/master/CONTRIBUTING.md#configure-eclipse)
- You need to add on top --> If your Eclipse is configured correctly, the imports will be added automatically.
   `import static name.abuchen.portfolio.util.TextUtil.trim;` <-- This already exists in Line 4.
   `import static name.abuchen.portfolio.util.TextUtil.replaceMultipleBlanks;`  <-- This must be added.

```suggestion
                            t.setNote(trim(replaceMultipleBlanks(v.get("note"))));
```

File: name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/santanderconsumerbank/Kontoauszug01.txt [hunk 4]
```diff
@@ -8,12 +8,13 @@ Kontoauszug Erstellungsdatum: 03.06.2023
...
-€ 0,00 € 38,98 € 40,64 € 1,66 € 6,63 € 1,66
+€ 0,00 € 37,98 € 40,64 € 2,66 € 6,63 € 1,66
 Buchungsdatum Valutadatum Betrag Saldo Buchungstext
 15.05.2023 15.05.2023 0,01 0,01 Einzahlung von Konto  AT123456789012345678 lautend auf Max Muster
 19.05.2023 19.05.2023 34,00 34,01 Einzahlung von Konto  AT123456789012345678 lautend auf Max Muster
-31.05.2023 31.05.2023 6,63 40,64 Zinsgutschrift Habenzinsen
-31.05.2023 31.05.2023 -1,66 38,98 Kapitalertragsteuer
+20.05.2023 20.05.2023 -1,00 33,01 Auszahlung auf Konto IBAN AT123456789012345678
+31.05.2023 31.05.2023 6,63 39,64 Zinsgutschrift Habenzinsen
+31.05.2023 31.05.2023 -1,66 37,98 Kapitalertragsteuer
```
Comment (Nirus2000) [inline name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/santanderconsumerbank/Kontoauszug01.txt:17]: _"The test files should not be modified beyond the anonymization."_
- see [Contributing rules --> Test cases](https://github.com/buchen/portfolio/blob/master/CONTRIBUTING.md#test-cases-1) .... Please note in future.

**Review by Nirus2000 — APPROVED** [review]

**Comment by Nirus2000** [comment]
In you last commit bae77bbc85e27d73de9740efa1b8f55a930f748f you need to fix this lines with this issues
- Improve regulare expression
- Missing "$" at the end of regulare expression and obsolet grouping brackets
- Remove obsolet mapping --> see https://github.com/buchen/portfolio/pull/3460#discussion_r1271538916
- and in the ".match("....") "... the regulare expression like my _suggested change_ 

1. https://github.com/buchen/portfolio/blob/befb01d306f07d0f469ce7c74eb879ef8d866126/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java#L189
2. https://github.com/buchen/portfolio/blob/befb01d306f07d0f469ce7c74eb879ef8d866126/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java#L212
3. https://github.com/buchen/portfolio/blob/befb01d306f07d0f469ce7c74eb879ef8d866126/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java#L237

and here like this comment --> https://github.com/buchen/portfolio/pull/3460#discussion_r1271539620
1. https://github.com/buchen/portfolio/blob/befb01d306f07d0f469ce7c74eb879ef8d866126/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java#L231
2. https://github.com/buchen/portfolio/blob/befb01d306f07d0f469ce7c74eb879ef8d866126/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java#L256

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java [hunk 5]
```diff
@@ -253,7 +253,32 @@ private void addAccountStatementTransaction()
...
+
+                        .section("date", "amount", "note")
+                        .match("^(?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) [\\d]{2}\\.[\\d]{2}\\.[\\d]{4} \\-(?<amount>[\\.,\\d]+) [\\.,\\d]+ Auszahlung (?<note>.*)$")
+                        .assign((t, v) -> {
+                            Map<String, String> context = type.getCurrentContext();
+
+                            t.setDateTime(asDate(v.get("date")));
+                            t.setAmount(asAmount(v.get("amount")));
+                            t.setCurrencyCode(asCurrencyCode(context.get("currency")));
+                            t.setNote(replaceMultipleBlanks(v.get("note")).trim());
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java:281]:
```suggestion
                            t.setNote(trim(replaceMultipleBlanks(v.get("note"))));
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java [hunk 6]
```diff
@@ -253,7 +253,32 @@ private void addAccountStatementTransaction()
                             t.setDateTime(asDate(v.get("date")));
                             t.setAmount(asAmount(v.get("amount")));
                             t.setCurrencyCode(asCurrencyCode(context.get("currency")));
-                            t.setNote(TextUtil.replaceMultipleBlanks(v.get("note")).trim());
+                            t.setNote(replaceMultipleBlanks(v.get("note")).trim());
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/SantanderConsumerBankPDFExtractor.java:256]:
```suggestion
                            t.setNote(trim(replaceMultipleBlanks(v.get("note"))));
```

**Review by Nirus2000 — APPROVED** [review]

**Comment by Nirus2000** [comment]
Hello @auchri 
The commit ID in my post https://github.com/buchen/portfolio/pull/3460#issuecomment-1646929439 was not quite correct.
Your commit with the outstanding issues, which I mentioned [here](https://github.com/buchen/portfolio/pull/3460#issuecomment-1646929439), is: https://github.com/buchen/portfolio/commit/f7dc7a0bb2a5ccf27aaf0f0897549b3841108db1

Regards
Alex

**Comment by buchen** [comment]
Thanks @auchri - looks nice and sweet to me.

**Review by Nirus2000 — APPROVED** [review]

**Comment by auchri** [comment]
I've resolved the merge conflicts

**Comment by buchen** [comment]
Thanks @auchri !!!

**Pull request merged**